Question board

moonhuicheol opened the pull request:
- 기존의 main에 머지했던 게시판 정규화는 적용하지 않았습니다.(git branch가 엉킬까봐..)
   PR통과후에 main과 머지한 후 적용하도록 하겠습니다.
- FE 분들과 수정했던 질문게시판 댓글부분은  답변을 달 튜터의 프로필사진을 나열하는 방식으로 만들기로했습니다.
   따라서 아직 프로필사진을 아직 구현하지 못해서 그 부분 제외하고 완료했습니다.

moonhuicheol committed:
questionBoard

moonhuicheol committed:
questionBaord 기능구현

moonhuicheol committed:
questionBoard.sql추가

zero2080 reviewed (commented):
### 1. 파일이나 함수 이름의 규칙을 일관되게 유지해주세요..
  어디서는 카멜케이스고...어디서는 캐밥이고...어디서는 스네이크고....어디서는 섞어 쓰고....
다른사람이 봤을때 운좋으면 아직 정립이 안됐다고 넘어갈텐데..
어떤사람은 어디서 복사해올때 파일통째로 가져왔다고 생각할 수도 있어요...

### 2. 스키마 키 설정해주세요.
  질문 게시판 테이블에 관계 설정이 안되어있습니다.

### 3. 로그인 시 가입한 이력이 없는 이메일로 로그인 시도시 서버가 종료됩니다.

### 4. `subject` 분류에 대해서 정규화 해주세요.
   - 서브젝트를 어떤 문자열로 올려도 올라갑니다.

### 5. `status` 의 용도나 사용법에대한 가이드가 없습니다.

### 6. `reIssueToken` 미들웨어가 정상적으로 작동하지 않습니다. 
  - access토큰만 전달하면 토큰이 정상이 아니더라도 미들웨어가 통과됩니다.

moonhuicheol committed:
questionBoard 1차수정

moonhuicheol commented:
5. status 부분은 추후에 채팅기능 개발할 시  질문게시판 filter에 해결, 미해결 상태를 넣기 위해 만들어 놓은 컬럼입니다.
채팅 개발할때 사용하려고 미리 만들어 놓았습니다.

moonhuicheol committed:
apiTest

moonhuicheol committed:
api 1차 최신화

zero2080 commented on index.js:
```diff
@@ -8,6 +8,7 @@ const port = 3000;
 const signup = require("./router/signup");
 const login = require("./router/login");
 const freeBoard = require("./router/freeBoard");
```
파일명은 `freeboard.js`입니다. 맞춰주세요.
AWS의 node환경은 파일의 대/소 문자도 구분합니다.

나머지 파일들도 require로 임포트하는 부분에서 대/소문자를 정확히 기입했는지 확인해 주세요.

zero2080 commented on middleware/authorization.js:
```diff
@@ -6,18 +6,19 @@ const bcrypt = require("bcrypt");
 
     if (!row) {
-        return res.status(400).json( {message: "해당 이메일이 없습니다."});
+        return res.status(400).json({ message: "해당 이메일이 없습니다." });
```
```suggestion
        return res.status(404).json({ message: "해당 이메일이 없습니다." });
```
찾으려는 데이터가 없을때는 404 NOT FOUND입니다.

zero2080 commented on middleware/authorization.js:
```diff
@@ -6,18 +6,19 @@ const bcrypt = require("bcrypt");
 
         if (!isPw) {
-            return res.status(400).json({ message: "잘못된 비밀번호 입니다."});
+            return res.status(400).json({ message: "잘못된 비밀번호 입니다." });
```
```suggestion
            return res.status(401).json({ message: "잘못된 비밀번호 입니다." });
```
400은 잘못된 요청을 보냈을때 사용합니다. 

**400 Bad Request**
> `Content-Type` 형식을 지원하지 않을때
> `Accept` 형식을 지원하지 않을때
> 필수 필드가 없을때
> 악의적인 요청이 감지 되었을때 등등...

401 Unauthorized
> 인증되지 않은 클라이언트 접근

zero2080 commented on middleware/authorization.js:
```diff
@@ -56,32 +57,40 @@ function generateRefreshToken(rowInfo) {
-    if (accessToken === "TokenExpiredError" && refreshToken === "TokenExpiredError") {
+    console.log("accessToken", accessToken);
+    console.log("refreshToken", refreshToken);
+    if ((accessToken === "TokenExpiredError") && (refreshToken === "TokenExpiredError")) {
```
```suggestion
    if ( !req.headers["authorization"] && !req.headers["refresh"] ){
      return res.status(403).json({message:"not authenticated"});
    }
    if (refreshToken === "TokenExpiredError") {
```

첫번째 조건의 추가로 두개의 헤더가 필수임을 명시 합니다.

두번째 조건으로 리프레시토큰이 만료됐는지 확인합니다.

accessToken은 refreshToken으로 생성할 수 있기때문에 refreshToken만 만료 확인을 하면됩니다.

zero2080 commented on middleware/authorization.js:
```diff
@@ -56,32 +57,40 @@ function generateRefreshToken(rowInfo) {
+    console.log("refreshToken", refreshToken);
+    if ((accessToken === "TokenExpiredError") && (refreshToken === "TokenExpiredError")) {
 
         return res.redirect("/auth/login");
```
```suggestion
        return res.status(401).json({message:"token expired"});
```

이 API를 호출하는 주체는 자바스크립트 단에서 이루어 집니다. 
그렇기때문에 리다이렉트는 정상적으로 처리 되지 않습니다.

그러므로 프론트에서 `refreshToken`이 만료되었음을 알려줘야 합니다.

만료응답을 받고, 프론트에서 로그인 페이지로 보낼겁니다.

물론 API문서에 해당내용이 포함되어있어야 합니다.

zero2080 commented on middleware/authorization.js:
```diff
@@ -56,32 +57,40 @@ function generateRefreshToken(rowInfo) {
     }
 
-    if (accessToken === "TokenExpiredError" && refreshToken) {
+    if ((accessToken === "TokenExpiredError") && refreshToken) {
```
```suggestion
    if (accessToken === "TokenExpiredError") {
```

기존의 조건문은 `refresh`헤더를 설정하지 않았을때 `verifyToken`메서드가 `JsonWebTokenError`을 반환합니다.
그렇기때문에 기존 조건문은 `true`가 됩니다.
기존 조건문이 성립되는 조건은 존재하지 않습니다. 
`refreshToken`은 에러 혹은 토큰 둘중하나이기때문에 
**`refreshToken`변수는 `falsy`가 절대 될 수 없습니다.**

변경한 조건문은 상단에서 두개의 초큰이 무조건 헤더에 포함되어있어야만하는 상황을 강제합니다.
때문에 `accessToken`의 만료 여부만 확인하면 됩니다.

zero2080 commented on middleware/authorization.js:
```diff
@@ -56,32 +57,40 @@ function generateRefreshToken(rowInfo) {
         });
 
-        return res.send({ accessToken: newAccessToken });
+        return res.status(400).json({ accessToken: newAccessToken });
```
```suggestion
        res.setHeader("Authorization",newAccessToken);
```

새로운 토큰을 발급하고 해당 토큰을 헤더에 담아서 다음 요청으로 넘깁니다.

zero2080 commented on middleware/authorization.js:
```diff
@@ -56,32 +57,40 @@ function generateRefreshToken(rowInfo) {
 
-    if (accessToken && refreshToken === "TokenExpiredError") {
+    if (accessToken && (refreshToken === "TokenExpiredError")) {
         const newRefreshToken = generateRefreshToken({
             id: accessToken.id,
             email: accessToken.email,
             status: accessToken.status
         });
 
-        return res.send({ refreshToken: newRefreshToken });
+        return res.status(400).json({ refreshToken: newRefreshToken });
     }
 
+    if (accessToken === "JsonWebTokenError") {
+        return res.status(400).json({ message: "accessToken이 존재하지 않습니다. "});
+    }
+
+    if (refreshToken === "JsonWebTokenError") {
+        return res.status(400).json({ message: "refreshToken이 존재하지 않습니다. "});
+    }
```
```suggestion
```
리프레시 토큰은 재발급 하지않습니다.
리프레시 토큰은 액세스 토큰을 발급해주기위한 토큰이기때문에 로그인을 통해서만 발급합니다.

zero2080 reviewed (changes requested):
1. 메서드, 클래스, 변수 이름이 변경되면 해당 기능을 사용하는 곳도 모두 바꿔주세요.
2. http status 상태에 대해서 숙지 하세요.

## PR올리기 전에 영향이 있는 기능은 무조건 테스트를 해야합니다.

moonhuicheol committed:
questionBoard 2차수정

moonhuicheol committed:
token 수정

zero2080 committed:
수정사항

 - 검증
   - `accessToken`과 `refreshToken` 유무 확인 조건문을 수정합니다.
     - `&&` > `||`
   - `reIssueToken` 메서드에서 토큰정책을 확인합니다.
   - 토큰이 재발행되면 response 헤더에 Authorization 키에 재발급된 토큰이 담깁니다.
   - 또한 다른 컨트롤러에서 다시 토큰을 해석하는 로직을 제거합니다. 대신 `reIssueToken` 미들웨어에서 request에 user정보를 저장합니다.
   - 이후 `reIssueToken`미들웨어를 거치는 모든 요청은 request에 user정보를 담게 됩니다.

zero2080 reviewed (approved):
너무 길어져서 직접 수정사항 커밋 했습니다.

커밋 메세지 확인해 주세요.
